### DISCUSSION TITLE
Purpose of module macro?

### DISCUSSION THREAD

**Discussion by VorpalBlade** [rune-rs/rune#718] [Q&A] [answered]
* There is a [module macro](https://docs.rs/rune/latest/rune/attr.module.html) that isn't really documented.
* It is [not used](https://rune-rs.github.io/book/external_types.html) when declaring a module in the book.
* It is [used](https://github.com/rune-rs/rune/blob/999eac032df1148baadde61fc412bebd23630b12/crates/rune-modules/src/json.rs) in at least some of the internal modules.

This leads to two questions: 
* If I want to declare a custom native module with a bunch of types and functions (that is not in the root namespace, but in a proper module), do I need/want that macro?
* What if I *do* want to put something in the root namespace?

**Comment by udoprog** [comment, marked as answer]
You don't need to use the macro. It's only used to define the name and documentation for the internal modules right now, something [you can do by hand with `Module::item_mut`](https://docs.rs/rune/latest/rune/struct.Module.html#method.item_mut).

If you want to put something in the root namespace, you use [`Module::new`](https://docs.rs/rune/latest/rune/struct.Module.html#method.new).
